quark/foreach.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 *--
+
+import os
+import subprocess
+from argparse import ArgumentParser
+
+from quark.utils import DirectoryContext
+
+from .subproject import Subproject
+
+
+def run():
+    parser = ArgumentParser(
+        description="""
+        Evaluates an arbitrary shell command in each submodule, skipping all
+        the svn's submodules.
+        The command has access to the variables $name, $sm_path, $displaypath,
+        $sha1, $toplevel, $rev, $version_control:
+        $name is the name of the submodule;
+        $sm_path is the path of the submodule relative to the superproject;
+        $displaypath is the path of the submodule relative to the root
+        directory;
+        $version_control is the version control used by the subproject
+        (git/svn);
+        $sha1 is the commit of the subproject ( empty string if it is a
+        svn repository );
+        $rev is the revision of the subproject ( empty string if it is a
+        git repository );
+        $toplevel is the absolute path to the top-level of the
+        immediate superproject.
+        """
+    )
+    parser.add_argument(
+        "-q", "--quiet",
+        action="store_true",
+        help="Only print error messages"
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        nargs="+",
+        help="The command that will be run for every dependency"
+    )
+
+    optlist = parser.parse_args()
+
+    root, modules = Subproject.create_dependency_tree(
+        os.getcwd(), update=False
+    )
+
+    for path, module in modules.items():
+        cmd = " ".join([
+            os.path.abspath(x) if os.path.exists(x) else x
+            for x in optlist.command[0].split()
+        ])
+
+        cmd_env = dict(os.environ)
+        cmd_env.update(module.get_env_variables(toplevel=os.getcwd()))
+
+        with DirectoryContext(module.directory):
+            output = subprocess.check_output(cmd, shell=True, env=cmd_env)
+
+            if not optlist.quiet:
+                print(output.decode("utf-8"))
```

Comment:
Capturing just to print is generally a bad idea, as it:

- consumes memory for the captured data, which may be a problem when running an arbitrary user command that may generate whatever quantity of data;
- messes up synchronization between `stdout` and `stderr` - you see `stderr` onscreen as it is generated, while you get `stdout` all at the end;
- if the command hangs, you don't see any output;
- with Python 3, you enter the usual world of encoding-related pain, as `print` expects `str` but `check_output` correctly returns `byte`; the correct thing in this regard would be to do `sys.stdout.buffer.write(output)`; but again, you don't want to do this.

If this is functionality that _we actually need_, the correct thing is to pass [`stdout = subprocess.DEVNULL`](https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL) (and possibly the same for `stderr`) if we want to mute the command output.

Still, this it must be noted that this is _not_ what `git submodule foreach` does; the `--quiet` option applies to messages _that git itself_ generates, not to the command, which is executed as-is, without touching its stdout/stderr. IMO the `git submodule foreach` approach is more correct - if the user wants its command to be quiet, they will just add a `> /dev/null` at its end.

So:

- if we want to be coherent with `git submodule foreach`, lines 61:64 would be replaced just by `subprocess.check_call(cmd, shell=True, env=cmd_env)`;
- if we are really attached to this "quiet the executed command" idea, then they should be replaced by `subprocess.check_call(cmd, shell=True, env=cmd_env, stdout = subprocess.DEVNULL if optlist.quiet else None)`.

Personally I'm voting for the first option.

Reply:
Changed in https://github.com/comelz/quark/pull/27/commits/4c36274c4eb79ecbcc46b7adc38fc5f5c381c0c5

The option `--quiet` now works like `git submodule -q foreach ...` ( changed in https://github.com/comelz/quark/pull/27/commits/9993e95e1c2c954fd405da1d3facfe32d761f6f5 )